Candidate RP2350 upstream+ branch

P33M opened the pull request:
I've already bumped master (which tracks upstream) to a recent post-0.12 commit. 0.13 isn't coming anytime soon, which is unfortunate as there are 996 commits on top.

This adds @tom-van's risc-V changes and re-added downstream warts to maintain compatibility with our documentation.

Some intensive testing is required.

timsifive committed:
Make watchpoint.unique_id a uint32_t

Now it matches breakpoint.unique_id.

Change-Id: I06f24b2cede2ee56bdeac8666b5235f923b18659

P33M committed:
tcl/interface: replicate downstream raspberrypi-swd.cfg file

Just include the upstream config. Note that the pin connections have
changed, with good reason

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
interface: raspberrypi-swd: add override for downstream pin connections

Existing setup guides use the old layout.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

kr-sc committed:
target/riscv: Improve riscv controls that manage the set of available triggers for watchpoints

Add more debug messages connected with triggers.
Update names for internal flags to make them more clarified.

Change-Id: I5642346ce4a1e9bf79b22cdbf36bd757a7beffa8
Signed-off-by: Kirill Radkin <kirill.radkin@syntacore.com>

en-sc committed:
target/riscv: set `state` and `debug_reason` in `riscv_halt_go_all_harts()`

If targets are in a halt group, and a target in the group reaches a
breakpoint, the target's state was able to remain `TARGET_RUNNING`.

Addresses issue #1010

Change-Id: I734bc6da71d289c4d05b417c8bf67a7d1a56574f
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

tom-van committed:
target/riscv: add msleep register

RP2350-A0 requires fiddling with sleep config

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I6edde38508f2cab41529ce622fde4f666627e6f5

timsifive committed:
target, flash/nor: main tree part of `riscv repeat_read` command

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I853bcdfc5bf5863f12a56b8bf4e6447f0a2f6e38

timsifive committed:
helper: add base64

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I24412eda300f35e23c5ce669616d37de0da755a1

timsifive committed:
target/riscv: unofficial bulk update

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I35a3662d9e26f85aa49471737e57377029ace4c0

tom-van committed:
target/riscv: respect DM_DMSTATUS_ALLUNAVAIL bit during examine

Do not throw nonsense errors if a hart has DM_DMSTATUS_ALLUNAVAIL
set. Postpone examination till DM_DMSTATUS_ALLUNAVAIL is cleared.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I94ccf5cc80e104abb606c7c3798ff70bd8f8e707

JanMatCodasip committed:
riscv/program: Removed dead code for restoring register values

Function riscv_program_exec() contains code for restoring
of register values after progbuf execution. This code is
not used anymore by current OpenOCD, and hence removed.

Related discussion can be found under:
https://github.com/riscv/riscv-openocd/issues/982

Change-Id: I4c79bec081522b6fc0d16367cef51ed19a131962
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Cosmetic cleanup of dm_*() calls in riscv-013.c

Cleanup, non-functional changes:
- Replaced one call of low-level function dm_op()
  by high-level dm_read().
- Made sure that truncation of values passed to dm_*
  is explicit. (Added explicit casts.)

Change-Id: I1d1b2f29a822b6841373f3313de2b1e96f514116
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Fixes and cleanup in riscv batch and related functions

Fixes:

- Data types of address & data parameters in riscv_batch_add_*()
  and riscv*_fill_dm*() changed to uint64_t and uint32_t.

- Corrected the comparison in riscv_batch_full().

- Corrected assertions in riscv_batch_get_dmi_read_op()
  and riscv_batch_get_dmi_read_data().

Cleanup:

- Simplified calloc() fail handling in riscv_batch_alloc().

- Added explicit NULL assignments in riscv_batch_alloc()
  for clarity and readability. Don't rely on calloc().

- Removed suffix `_u64` from riscv_*_fill_dm*() since it
  does not have any meaning.

- Renamed *dmi_write_u64_bits() to *get_dmi_scan_length()
  which better describes its purpose.

Change-Id: Id70e5968528d64b2ee5476f1c00e08459a1e291d
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Fixes of review findings

Change-Id: Ie9889d995d7b2a6e458ad5f66cc3d990888f54ec
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

tom-van committed:
target/arm_adi_v5: do not invalidate DAP cache before dp_init

At least select_dpbanksel_valid flag set by DP connect
should be preserved to prevent unwanted write to SELECT register.

Move dap_invalidate_cache() before connect wherever applicable.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I238c0a546b7451e4185a5895b777b8282ad996eb

tom-van committed:
target/riscv: read_memory_bus_v1() decoupled from JTAG

A rewrite from scratch of read_memory_bus_v1()
- uses riscv_batch instead of direct calls to JTAG scans
- adjusts to target device memory speed by inserting
  sbcs reads - not as smooth timing as JTAG runtest
- does not use reading sbaddress to get the exact point
  of busy error - reads sbcs more frequently
Can work over virtually any transport.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I7fb559941988b0ac969f93529cfe5f9a3faea003

tom-van committed:
target/riscv: DM access on a DAP

Modify riscv target to optionally accept -dap and -ap-num options
instead of -chain-position
The selected AP is used as an alternative DMI.
DTM version is fixed to 1 because DTM control scan is not available.

Basic DM access works.

riscv_batch works without a configurable delay. Also an optional
read back in riscv_batch_add_dm_write() is not inplemented
- the feature is used by memory sampling only.

read_memory_bus_v1() does not work as it uses jtag_add_runtest()
directly. Will be addressed in the following patch.

Checkpatch-ignore: EMBEDDED_FUNCTION_NAME
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I5399207d897d93419198c40b521e4464e04ff1a6

tom-van committed:
target/riscv: free dm and target_list structures

Fix memory leak on exit.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I6a89ac0c93b11da35b90eec3abcc5b6fd5d1be68

tom-van committed:
target/adi_v5_swd: make SWD connect DP lockup proof

An AP or a block/bus connected to it can be unpowered or
without a clock signal. In such situation DP locks up
on access to the AP or the non-functional block/bus.
In such state DP responses to IDR reads, CTRL_STAT reads
and ABORT writes by OK ack to comply with ADIv5.2 or v6
reference. All other accesses are refused with SWD WAIT.

Modify the (re)connect loops to respect SWD WAIT and keep
retrying until timeout and then issue DAPABORT to
revive the stalled DP.

Add direct write to DP SELECT register to swd_connect_single()
to detect stalled DP state and retry the connect loop.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I9d84792b6da08b967864c1ea4624a4fe8e9a006f

tom-van committed:
target/adi_v5_swd: switch to DORMANT to revive stalled DP

On several devices it was observed that DAPABORT bit
in ABORT DP register does not help to revive DP to functional
state. On the other hand, switching the DPv2 or v3
to DORMANT state reliably released the lock up and DP
can continue communicating to accessible APs.
Although it is not mentioned in ADI reference it
looks naturally that the DP logic gets reset after unpowering
or deep sleep in the DORMANT state.

Alternate issuing DAPABORT and switching to DORMANT state
after (re)connect timeout is over.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I9afdf971aa4b849ea48fdba40b748d0c7bd65130

roliver-rpi committed:
flash/nor/spi: Add Winbond w25q128jw

Signed-off-by: Richard Oliver <richard.oliver@raspberrypi.com>

Wren6991 committed:
target/riscv: Fix bug in riscv_openocd_poll():

This code forced all harts to halt when they reported a mixture of
halted and non-halted status. This breaks long-running algorithms run
on a single core under SMP, because the first poll will force the core
executing the algorithm to halt.

Fix by adding an exception when all running cores are running under
debugger control (target->status == 4). Do not force harts to halt in
this case.

Change-Id: I940172926ed9b80b95891d8aecbd10897e148585

tom-van committed:
target/riscv: do not set DTM_DTMCS_VERSION_UNKNOWN on examine fail

Proper version is necessary for target deinit or future
re-examinations. Poisoning the version number causes memory leak.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: Ie63ae83626c3fd7498669cb2898309f1c5d4d112

tom-van committed:
target, riscv: add target events to track availability changes

Emit 'become-unavailable' and 'become-available' events.

TODO: doc!

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I5a51f4cfbab26ce07ca03af7f95b71191a79bd1b

tom-van committed:
target/riscv: invalidate cached progbuf in unavailable state

Doesn't seem to be specified in RISC-V External Debug Support spec,
however at least the RISC-V cores of RP2350 clobber progbuf when
they are switched to the 'hart unavailable' state.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: Iea6e4f41d6430c1126e97d0f4f8161b441368510

tom-van committed:
target, riscv: examine the target when it becomes available

Poll the target with TARGET_UNAVAILABLE state even if not examined.
Emit TARGET_EVENT_BECOME_AVAILABLE which should be configured to
issue Tcl command 'arp_examine' if needed.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I0476dafb712b99424f6ce011505d71ae9eb81f5e

tom-van committed:
target/target: add Tcl event to check target availability

Extend target_handle_event() to target_handle_event_ret() which
could return a string from Tcl command.
Introduce target event check-availability and call it before poll
and before examine.

TODO: doc!
Add checking to ocd_process_reset_inner?

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I285fe177bc80f70b3bcaafc08521441582aca2b6

tom-van committed:
target/riscv: fix write_memory_bus_v1() hang when hart is unavailable

Check DM status when we get DM_SBCS_SBBUSY or DM_SBCS_SBBUSYERROR.

Anyway the endless retrying deserves an operation timeout.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Change-Id: I7dcd40d4cafb570ba6f0f794c805d7641351a8ba

tom-van committed:
Revert "target: riscv: convert 'unsigned' to 'unsigned int'"

This reverts commit 134e56338d365319380092a7bc5becf82c6d47f2.

tom-van committed:
Revert "target: riscv: remove non-trivial 'unsigned' cast"

This reverts commit fec3b224214e3784b0c00970d2421212402da880.

tom-van committed:
Revert "target: riscv: convert 'unsigned' to 'unsigned int'"

This reverts commit 957eb741a0980408fe4d0682fccb99a183f90998.

tom-van committed:
tcl/target/rp2350: add architecture switching

Tcl variable USE_CORE selects what cores are debugged.
If both cores of one slot are selected, OpenOCD chooses the active one.

Change-Id: I5d5fd5544d5996b3baa2488d17cc26156fd03268
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

P33M committed:
tcl/target: add rp2040-rescue.cfg

For compatibility with Raspberry Pi documentation.
This invokes the base rp2040.cfg with the RESCUE flag set.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
tcl/target: add rp2350-rescue.cfg

For compatibility with Raspberry Pi documentation.
This invokes the base rp2350.cfg with the RESCUE flag set.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

tom-van commented:
@P33M , could you please submit https://github.com/raspberrypi/openocd/pull/132/commits/7242458c528ff71a38be15f228653e30295d2fc4 and https://github.com/raspberrypi/openocd/pull/132/commits/01508397399a1849ed2711d54b1830b7b5c6f887 to the upstream https://review.openocd.org ?
I think we can merge them soon. Thanks!

P33M commented:
That is the general idea.

Note that this PR is primarily for discussion - I'll push the branch directly in this repo once we bless it as good.

lurch commented on tcl/target/rp2040.cfg:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0-or-later
 
 # RP2040 is a microcontroller with dual Cortex-M0+ core.
-# https://www.raspberrypi.com/documentation/microcontrollers/rp2040.html
```
The RP2040 "documentation URL" (as opposed to the "product URL") would seem to be https://www.raspberrypi.com/documentation/microcontrollers/silicon.html#rp2040

lurch commented on tcl/target/rp2350.cfg:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0-or-later
 
 # RP2350 is a microcontroller with dual Cortex-M33 cores or dual Hazard3 RISC-V cores.
-# https://www.raspberrypi.com/documentation/microcontrollers/rp2350.html
```
The RP2350 "documentation URL" (as opposed to the "product URL") would seem to be https://www.raspberrypi.com/documentation/microcontrollers/silicon.html#rp2350

lurch commented:
ping also @will-v-pi in case he has any opinions from the VSCode-extension side of things.

P33M commented:
@tom-van our pico-sdk-tools which packages openOCD for Windows needs to use the (now deprecated) jimtcl submodule as no MSYS2 package for jimtcl is available. For the next release, which will presumably remove jimtcl as a submodule entirely what is the supported method for building a Windows binary?

P33M committed:
tcl/target: add rp2350 debug key loader

OTP can be used to lock down debug with a per-device 128-bit challenge.

Add a tcl helper function to load the key 1 bit at a time into the RP-AP.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
tcl/target: add rp2350-riscv.cfg

For compatibility with Raspberry Pi documentation.
This invokes the base rp2350.cfg and probes both Hazard3 cores.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

tom-van commented:
> @tom-van our pico-sdk-tools which packages openOCD for Windows needs to use the (now deprecated) jimtcl submodule as no MSYS2 package for jimtcl is available. For the next release, which will presumably remove jimtcl as a submodule entirely, what is the supported method for building a Windows binary?

Not going to happen in upcoming 1.0, at first in 1.1
This question is more for MSYS2 packagers. I assume they will keep the OpenOCD package recipe functional, probably by adding the jimtcl library package to satisfy the new dependency.

P33M committed:
jtag/drivers/cmsis-dap: add hack for old and broken Pi Debugprobe firmware

Older firmware versions declare a max packet_count of 2 but fail to handle
pipelined packets. In older openOCD versions this was not an issue as
libusb calls were synchronous, but performance enhancements expose the bug.

The firmware version is reported in the device descriptor so a workable
kludge is to filter on VID:PID at the enumeration stage.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
tcl/target/rp2350: define commands before use

Using the Rescue reset produces runtime errors such as:

Error: [rp2350.cm1] Execution of event check-availability failed:
/usr/local/bin/../share/openocd/scripts/target/rp2350.cfg:125: Error: invalid command name "_cm_available_examine"

Move declarations ahead of target configuration routines.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

tom-van committed:
target/cortex_m: avoid poll() if TARGET_UNAVAILABLE

cortex_m has no generic flags to poll core availability.
Availability must be checked in TARGET_EVENT_CHECK_AVAILABILITY
Tcl event.

Check target->state early in poll() and return ERROR_OK
in case of TARGET_UNAVAILABLE

Change-Id: I5683af20ef64f71f7bdfd5a527bcf80f4f964c43
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
Please add https://github.com/tom-van/openocd-rp2350-riscv/commit/c983fcfaa8a4af438dfddb825efe50c5dd6f3f4d
It fixes OpenOCD crash if cm33 cores has not been available since OpenOCD start and then debug connection stops working (unplug)

BTW: Using `rescue_reset` from OpenOCD command line brings back cm33 cores without unplug and user may continue loading the new app in gdb or OpenOCD (not sure if applicable in vscode)

P33M commented:
No crashes observed since.

Aside: we may have to kludge the "rp2040" branch to point to master, as old versions of the getting-started guide (and random places on the internet) have setup instructions that use the line `git clone https://github.com/raspberrypi/openocd.git --branch rp2040 --recursive --depth=1` rather than the default.

will-v-pi commented:
I've noticed a build failure on Arm MacOS (sorry I didn't spot it earlier):
```
src/jtag/drivers/amt_jtagaccel.c:16:10: fatal error: 'linux/parport.h' file not found
#include <linux/parport.h>
```
Passing `--disable-parport-ppdev` to configure didn't fix it, it gave the warning `configure: WARNING: --disable-parport-ppdev is not supported by the host CPU`

will-v-pi commented:
Applying that commit to this branch doesn't fix the issue for me either - I'm back to the original `linux/parport.h` error. Is there a configure switch to completely disable the parport?

tom-van commented:
Did you re-run ./bootstrap?

tom-van commented:
Yes you're right, I selected the wrong one from two patches. As they both fixes some missing headers, please apply both

will-v-pi commented:
Probably best for @P33M to apply both patches directly to this branch (https://github.com/openocd-org/openocd/commit/1347b693a508179db5b19279cd6a37753fe0dfd5 and https://github.com/openocd-org/openocd/commit/4732e40637682a805956fd99832c151a9d4a691e)

will-v-pi commented:
Tested on MacOS and it's working fine with debugprobe, although I did run into a `jimtcl` annoyance - I thought it was supposed to be compiled in statically into `openocd`, but on MacOS it was throwing an error when I didn't have the `libjim.0.83.dylib` file installed - workaround is simple enough, so not an issue

P33M commented:
That was a slightly janky rebase due to the "unofficial bulk update" commit, but I believe it's correct.

I think one more round of builds and a once-over smoke test and we can call this done?

will-v-pi commented:
That built fine, but I'm getting an error when trying to debug on a Pico 2 that is in BOOTSEL mode. I've also tested yesterday's build and got the same error so this isn't related to the new commits, but I don't get the error with the [sdk-2.0.0 branch](https://github.com/raspberrypi/openocd/tree/sdk-2.0.0).
```
Open On-Chip Debugger 0.12.0+dev-ge8b8aa3 (2025-06-13-13:27)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
DEPRECATED! use 'gdb port', not 'gdb_port'
DEPRECATED! use 'tcl port' not 'tcl_port'
DEPRECATED! use 'telnet port', not 'telnet_port'
CDLiveWatchSetup
Info : [rp2350.cm0] Hardware thread awareness created
Info : [rp2350.cm1] Hardware thread awareness created
ocd_process_reset_inner
adapter speed: 5000 kHz
Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E6633861A319A238
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 5000 kHz
Info : SWD DPIDR 0x4c013477
Info : [rp2350.cm0] Cortex-M33 r1p0 processor detected
Info : [rp2350.cm0] target has 8 breakpoints, 4 watchpoints
Info : [rp2350.cm0] Examination succeed
Info : [rp2350.cm1] Cortex-M33 r1p0 processor detected
Info : [rp2350.cm1] target has 8 breakpoints, 4 watchpoints
Info : [rp2350.cm1] Examination succeed
Info : [rp2350.cm0] starting gdb server on 50000
Info : Listening on port 50000 for gdb connections
Info : accepting 'gdb' connection on tcp/50000
[rp2350.cm0] halted due to breakpoint, current mode: Handler HardFault
xPSR: 0x01000003 pc: 0x000002f8 msp: 0x400e0308
[rp2350.cm1] halted due to debug-request, current mode: Thread 
xPSR: 0x09000000 pc: 0x000000da msp: 0xf0000000
Error: Failed to invoke rcp_init
Error: Failed to init Arm core 0 before ROM call
Info : no SDFP found
Error: RP2350 rev 2, QSPI Flash id = 0x000000 not recognised
Error: auto_probe failed
Error: Connect failed. Consider setting up a gdb-attach event for the target to prepare target for GDB connect, or use 'gdb_memory_map disable'.
Error: attempted 'gdb' connection rejected
```

I don't get that error on Risc-V, so this isn't too big a problem for me, as the main use case I have for debugging starting in BOOTSEL mode is to get the chip in Risc-V mode with `picotool reboot -c riscv -u` before debugging a Risc-V binary - but this might need documenting if it can't be fixed to prevent confusion.

will-v-pi commented:
Separate issue - I wanted to have a go with debug keys but got the error
```
target/rp2350.cfg:237: Error: can't read "__CHIPNAME": no such variable
```
It seems half the `_CHIPNAME` have one underscore and half have two underscores, which causes that error (I think it's meant to be one underscore)

Edit: looking further, I realise `__CHIPNAME` is local and `_CHIPNAME` is global, so `send_dbgkey` just needs a
```
if { $__CHIPNAME == "" } {
	global _CHIPNAME
	set __CHIPNAME $_CHIPNAME
}
```
like `rescue_reset`

tom-van commented:
> That built fine, but I'm getting an error when trying to debug on a Pico 2 that is in BOOTSEL mode.
> 
> ```
> ...
> [rp2350.cm0] halted due to breakpoint, current mode: Handler HardFault
> xPSR: 0x01000003 pc: 0x000002f8 msp: 0x400e0308
> [rp2350.cm1] halted due to debug-request, current mode: Thread 
> xPSR: 0x09000000 pc: 0x000000da msp: 0xf0000000
> Error: Failed to invoke rcp_init
> Error: Failed to init Arm core 0 before ROM call
> Info : no SDFP found
> Error: RP2350 rev 2, QSPI Flash id = 0x000000 not recognised
> ```
> 
> I don't get that error on Risc-V, so this isn't too big a problem for me, as the main use case I have for debugging starting in BOOTSEL mode is to get the chip in Risc-V mode with `picotool reboot -c riscv -u` before debugging a Risc-V binary - but this might need documenting if it can't be fixed to prevent confusion.

Yes, it's annoying for gdb user. If you have OpenOCD telnet cmd line available issuing `reset init` easily solves the problem with rcp_init, calls ROM API function FC and flash probe starts working. But we discourage use of any reset in `gdb_attach` event to preserve debug status. We should investigate why rcp_init fails in BOOTSEL. @Wren6991 any ideas?

tom-van commented:
It looks as OpenOCD problem in disabling interrupts during debug execution on ARMv8-M. We use PRIMASK register to do so but ARMv8-M with security extension has the secure and non secure PRIMASK registers. 

Edit: Not as easy as that. Setting PRIMASK_S and PRIMASK_NS does not help.
Cortex-M33 hard faults on the first instruction after setting bit CDS of DSCSR no matter how PRIMASK or C_MASKINTS is set.

SHCSR 0x4: HardFault exception active for the selected Security state.
HFSR 0x40000000: Forced. Indicates that a fault with configurable priority has been escalated to a HardFault exception, because
it could not be made active, because of priority, or because it was disabled.
MMFAR, BFAR, SFAR: 0x400e031c in the middle of msp_s stacking frame

I have no idea why it happens

tom-van committed:
target/cortex_m: fix debug reason after reset halt

[1] removed target_halt() from cortex_m_assert_reset()
It broke debug_reason tracking and the previous reason
was shown after reset halt.

Set debug_reason to DBG_REASON_DBGRQ during reset halt
preparation.

Fixes: [1] commit 226085065bdf ("target/cortex_m: drop useless target_halt() call")
Reported-by: Marc Schink <dev@zapb.de>
Change-Id: I685618ed158abde11f6e00eeeee1dfa8ed90952d
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
@kilograham commented in mail:

> In BOOTSEL mode:
> RAM is set to NS accessible in the SAU (region 2) - this is probably fine
> RAM is set to XN via MPU - this is a problem
> Probably worth disabling/enabling MPU around the algorithm to see if that helps

Device rebooted to BOOTSEL from OpenOCD and halted:
```
reset init
rp2xxx rom_api_call RB 2
resume
sleep 100
halt
```
Forced to secure mode setting CDS in DSCSR
```
mww 0xE000EE08 0x10000
```
Actually the secure MPU looks disabled (MPU_TYPE and MPU_CTRL)
```
> mdw 0xE000ED90 2
0xe000ed90: 00000800 00000000
```
However SAU is enabled (SAU_CTRL and SAU_TYPE):
```
> mdw 0xE000EDd0 2
0xe000edd0: 00000001 00000008
```
and region 2 has bit ENABLE = 1 which means SAU region is disabled (horrible bit naming) so RAM is NS
```
> mww 0xE000EDd8 2
> mdw 0xE000EDd8 3
0xe000edd8: 00000002 20000000 20081fe1
```
Setting the ENABLE bit to 0 - SAU region is enabled - solves the HardFault problem in rcp_init:
```
mww 0xE000EDe0 0x20081fe0
> flash probe 0
RP2350 rev 2, QSPI Flash win w25q128fv/jv id = 0x1840ef size = 16384 KiB in 4096 sectors
flash 'rp2xxx' found at 0x10000000
```

Not sure how to make an universal fix independent of particular SAU region use (e.g. if SAU set differently in application code)

kilograham commented:
the secure MPU is disabled on A2 but is still enabled on A4

tom-van commented:
On A2 simply switching SAU off works too.

tom-van commented:
I realised another issue: if the core was in NS mode and `rp2350_init_arm_core0()` forces secure mode during flash probe in gdb attach, it's not possible to resume the original app code. We should restore the original CDS bit value and also SAU_CTRL and MPU_CTRL if they were switched off.

kilograham commented:
Does it not make sense for cortex-m to save state and disable SAU and MPU in general before running algorithm then restore after in general?

kilograham commented:
On a side note, I’m glad the bootrom is making it hard to run arbitrary code while in  BOOTSEL mode :-)

P33M commented:
This noddy patch makes rcp_init run OK, target stops in the ROM and GDB doesn't complain too much:
```
diff --git a/src/flash/nor/rp2xxx.c b/src/flash/nor/rp2xxx.c
index 85c591104..0a6da0447 100644
--- a/src/flash/nor/rp2xxx.c
+++ b/src/flash/nor/rp2xxx.c
@@ -647,6 +647,15 @@ static int rp2350_init_arm_core0(struct target *target, struct rp2xxx_flash_bank
                }
        }

+       /* Yolo smash the MPU enable */
+       retval = target_write_u32(target, 0xE000ED94, 0);
+       if (retval != ERROR_OK)
+               LOG_ERROR("RP2350 mpu wiggle wrong ");
+
+       retval = target_write_u32(target, 0xE000EDD0, 0);
+       if (retval != ERROR_OK)
+               LOG_ERROR("RP2350 sau wiggle wrong ");
+
        if (!priv->stack) {
                LOG_ERROR("No stack for flash programming code");
                return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
diff --git a/tcl/target/rp2350.cfg b/tcl/target/rp2350.cfg
index 72c382922..e220de58c 100644
--- a/tcl/target/rp2350.cfg
+++ b/tcl/target/rp2350.cfg
```

However these registers are a) common to all cortex-M33 implementations b) independently optional c) have no effect if Secure invasive debug is disabled in OTP (or other platform-specific method).

tom-van commented:
But creates a hidden bomb (and difficult to find too) for anybody who wants to continue debugging after on-the-fly gdb attach.
Tried with BOOTSEL:
```
(gdb) target remote :3333
Remote debugging using :3333
0x00005e74 in ?? ()
(gdb) c
Continuing.
 
Program received signal SIGTRAP, Trace/breakpoint trap.
0x000002f8 in ?? ()
```

@kilograham:
> Does it not make sense for cortex-m to save state and disable SAU and MPU in general before running algorithm then restore after in general?

Good point. However other flash operations (erase/write etc...) does not need to restore system to the original state as continuation after re-flash is hardly possible. And the wast majority of flash drivers do probe without running any target algo.
So RP2350 flash is currently the only driver requiring security switch and restore. I admit there could be some more problematic flash drivers if an optional external QSPI is combined with secure-NS user app. I'm not aware of another callable ROM which can drop to NS.
Also RP2350 would need to save and restore ACCESSCTRL which is reset in both ARM and RISC-V modes.

P33M commented:
I think that for this target downstream-ish release of openOCD it's out of scope to handle all the corner cases with NS app, NS debug, NSC ROM, OTP disables etc. Unpicking those interactions is going to be a fair bit of work.

If we assume that a debug session and target algorithm / loaded app always runs in Secure mode, then the ACCESSCTRL/MPU/SAU disables can be unconditional. 

Edit: so this implies that you can't debug BOOTSEL, but are free to program flash / load a RAM binary.

will-v-pi commented:
Would it be possible to just call `FUNC_FLASH_EXIT_XIP` if you've started debugging an A2 chip and you're in NSBOOT? This is the bootrom bug on A2 (errata E10) which is causing this to be fatal, rather than just getting a failed rcp_init followed by a successful flash probe & subsequent debugging.

tom-van commented:
> Edit: so this implies that you can't debug BOOTSEL, but are free to program flash / load a RAM binary.

Don't take me wrong, I don't expect normal users want to debug BOOTSEL.
Imagine someone debugs an app with both S and NS modes. Attaching gdb to a running app doesn't seem to me as a corner case. If he does so when app is in NS everything goes wrong, no clue why.
Please at least add a bold warning that debugging can't be resumed after NS -> S switch.

> Would it be possible to just call `FUNC_FLASH_EXIT_XIP` if you've started debugging an A2 chip and you're in NSBOOT?

In RP2350 datasheet the function is marked as callable in Arm-S and RISC-V modes. Do I miss something?

will-v-pi commented:
> > Would it be possible to just call `FUNC_FLASH_EXIT_XIP` if you've started debugging an A2 chip and you're in NSBOOT?
> 
> In RP2350 datasheet the function is marked as callable in Arm-S and RISC-V modes. Do I miss something?

NSBOOT can call it through a secure call, which is made available for NSBOOT but not other NS code, so that's probably not the way to go. I was more thinking if you're already able to force the processor into secure mode, then it can be called from Arm-S, but I might be misunderstanding how calling ROM APIs from OpenOCD works.

P33M commented:
Would it be less unpalatable if the modifications to DSCSR (and MPU/SAU) are undone in cleanup_after_raw_flash_cmd? That seems to work for me - no stopping in a hardfault handler, though I do randomly get NMIs if I don't hit continue in GDB fast enough.

kilograham commented:
one side note, which may or may not be helpful, is that when debugging during the boot path itself, or during BOOTSEL mode, the RCP _will_ have already been initialized (or you could be debugging before that is needed)...

This means you could fix the issue (for BOOTSEL mode) by just fixing the flash probe. Note that attaching the debugger in BOOTSEL mode is something I do a lot, as it is a trivial way to get back to a debuggable system when you have done something to the clocks or whatever, and you can no longer attach the debugger (or if for whatever reason the binary on your device borks the debugger)

But yes, I agree, in general that being able to debug code when an MPU/SAU is in place is good

P33M commented:
The RCP algorithm code checks for re-entrancy and doesn't try to re-seed the canaries or sev() in that case. The error message was originally because the algorithm can't be run in the first place (RAM is set as XN etc).

Is it possible to reset the cores from the debugger without resetting the RCP, so the ROM runs and explodes when it tries to seed them?

tom-van commented:
Not sure, but on A2 we need to call jump_connect_internal_flash (FC) before probing, so we need to go secure anyway.

I'm working on support routines `cortex_m_set_secure()` and `cortex_m_security_restore()` to keep layering correct. They seem to work on A2, I have to add MPU switching. Please wait till tomorrow, I have to go now.

tom-van committed:
target/cortex_m: introduce security manipulation routines

Running target algorithms on ARMv8M may require core in secure
mode with SAU and MPU off (as set after reset).

cortex_m_set_secure() forces this mode with optional save of
the previous state.

cortex_m_security_restore() restores previously saved state.

Change-Id: Ia71826db47ee7b0557eaffd55244ce13eacbcb4b
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van committed:
flash/nor/rp2xxx: save security state over target algo

RP2040 and RP2350 flash driver runs a ROM API target algorithm
in probe to setup QSPI command interface. The Cortex-M33 core
of RP2350 has to be in secure mode with SAU and MPU switched off
to ensure ROM API call working properly.

Especially after the flash probe (used in gdb-attach event)
we need to completely restore the original security state to allow
'resume' or gdb 'continue' without injecting strange errors
to application code.

Use cortex_m support to set secure mode and to restore it back.

Fixes: commit ea775d49fc71 ("flash/nor/rp2040: add RP2350 support")
Change-Id: I72096bfecbb45a8aa4d3a7a37ad140532b3b00b2
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
Please test
https://github.com/tom-van/openocd-rp2350-riscv/commit/55281d9ef4bd1b45bd4b645b0536bc1972eef3a5
and
https://github.com/tom-van/openocd-rp2350-riscv/commit/e3d44750d6384115b32e44471574c5024eb95900

ACCESSCTRL saving is not yet covered.

tom-van committed:
flash/nor/rp2xxx: save ACCESSCTRL over ROM API calls

Especially after the flash probe (used in gdb-attach event)
we need to completely restore the original security state to allow
'resume' or gdb 'continue' without injecting strange errors
to application code.

Save all ACCESSCTRL registers potentially changed by triggering CFGRESET.
Restore them at cleanup.

Fixes: commit ea775d49fc71 ("flash/nor/rp2040: add RP2350 support")
Change-Id: I964886d5b1d0269497c343811ee4dcd5c31953db
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
ACCESSCTRL save and restore:
https://github.com/tom-van/openocd-rp2350-riscv/commit/dad9e7f4039ef962cfc336e2e4285f71e0cbeb37

will-v-pi commented:
Those changes have fixed the issue for me

P33M commented:
If I attach GDB under BOOTSEL and am quick about resuming the core, the "files" on the mass storage gadget are still readable.
